src/julia_threads.h

```diff
@@ -74,6 +77,17 @@ typedef struct _jl_tls_states_t {
 #  define JL_CPU_WAKE_NOOP 1
 #endif
 
+// Copied from libuv. Add `JL_CONST_FUNC` so that the compiler
+// can optimize this better.
+static inline unsigned long JL_CONST_FUNC jl_thread_self(void)
+{
+#ifdef _OS_WINDOWS_
+    return (unsigned long)GetCurrentThreadId();
+#else
+    return (unsigned long)pthread_self();
+#endif
+}
```

Comment:
> Not all of pthreads is implemented in all UNIX-like systems, so unfortunately, it isn't just the Windows thread API and the pthreads API as in this case, but quite a mess of `#ifdefs`.

At least it is for this case. I'm not saying we should throw away pthread for everything. The lock API for example is still useful for a number of cases.

> I do think they should be collected in one place rather than having a mess of conditionals in a number of different places.

They are not. Other than what's in `threading.c` before, most (if not all) of the newly added threading related code are in `julia_thread.h` as inline functions or macros and they are as "collected in one place" as I can possibly make without breaking cross reference etc.